source/robot_model/src/Model.cpp

```diff
@@ -260,9 +262,138 @@ std::vector<state_representation::CartesianPose> Model::forward_geometry(const s
   return this->forward_geometry(joint_state, frame_ids);
 }
 
-state_representation::JointPositions Model::inverse_geometry(const state_representation::CartesianState&) const {
-  // TODO
-  return state_representation::JointPositions();
+Eigen::MatrixXd Model::cwln_weighted_matrix(const state_representation::JointPositions& joint_positions, const double& margin) {
+  Eigen::MatrixXd W_b = Eigen::MatrixXd::Identity(this->robot_model_.nq, this->robot_model_.nq);
+  for(int n=0; n < this->robot_model_.nq; n++){
+    double d=1;
+    if(joint_positions.get_positions()[n] < this->robot_model_.lowerPositionLimit[n]+margin){
+      d = (this->robot_model_.lowerPositionLimit[n]+margin - joint_positions.get_positions()[n])/margin;
+    }
+    else if(this->robot_model_.upperPositionLimit[n] - margin < joint_positions.get_positions()[n]){
+      d = (joint_positions.get_positions()[n] - (this->robot_model_.upperPositionLimit[n]-margin))/margin;
+    }
+    if((joint_positions.get_positions()[n] < this->robot_model_.lowerPositionLimit[n]) || (this->robot_model_.upperPositionLimit[n] < joint_positions.get_positions()[n])){
+      W_b(n,n) = 0;
+    }
+    else if((joint_positions.get_positions()[n] < this->robot_model_.lowerPositionLimit[n]+margin) || (this->robot_model_.upperPositionLimit[n] - margin < joint_positions.get_positions()[n])){
+      W_b(n,n) = -2*d*d*d + 3*d*d;
+    }
+    else{
+      W_b(n,n) = 1;
+    }      
+  }
```

Comment:
- `else` statements are to be continued in the same line `} else {`
- spaces around code flow statements (if / while / for) `for (...) {`
- spaces around operators (`n = 0`, `a / b`)
- column limit exceeded (lines are too long)

```suggestion
  for (int n = 0; n < this->robot_model_.nq; n++) {
    double d = 1;
    if (joint_positions.get_positions()[n] < this->robot_model_.lowerPositionLimit[n] + margin) {
      d = (this->robot_model_.lowerPositionLimit[n] + margin - joint_positions.get_positions()[n]) / margin;
    } else if (this->robot_model_.upperPositionLimit[n] - margin < joint_positions.get_positions()[n]) {
      d = (joint_positions.get_positions()[n] - (this->robot_model_.upperPositionLimit[n] - margin)) / margin;
    }
    if ((joint_positions.get_positions()[n] < this->robot_model_.lowerPositionLimit[n])
        || (this->robot_model_.upperPositionLimit[n] < joint_positions.get_positions()[n])) {
      W_b(n, n) = 0;
    } else if ((joint_positions.get_positions()[n] < this->robot_model_.lowerPositionLimit[n] + margin)
        || (this->robot_model_.upperPositionLimit[n] - margin < joint_positions.get_positions()[n])) {
      W_b(n, n) = (-2 * d * d * d) + (3 * d * d);
    } else {
      W_b(n, n) = 1;
    }
  }
```

Reply:
As a subjective comment, it's a pretty dense block already so "good readability" is going to be hard to achieve. But, I think you are doing more conditional statements than necessary (you have 6 conditions, if I count the `if (A || B)` as two conditions.

You could reformat it a bit as follows:

```
    double d = 1;
    W_b(n, n) = 1;
    if (joint_positions.get_positions()[n] < this->robot_model_.lowerPositionLimit[n] + margin) {
      d = (this->robot_model_.lowerPositionLimit[n] + margin - joint_positions.get_positions()[n]) / margin;
      if (joint_positions.get_positions()[n] < this->robot_model_.lowerPositionLimit[n]) {
        W_b(n, n) = 0;
      } else {
        W_b(n, n) = -2 * d * d * d + 3 * d * d;
      }
    } else if (this->robot_model_.upperPositionLimit[n] - margin < joint_positions.get_positions()[n]) {
      d = (joint_positions.get_positions()[n] - (this->robot_model_.upperPositionLimit[n] - margin)) / margin;
      if (this->robot_model_.upperPositionLimit[n] < joint_positions.get_positions()[n]) {
        W_b(n, n) = 0;
      } else {
        W_b(n, n) = -2 * d * d * d + 3 * d * d;
      }
    }
```